src/iop/colorbalance.c

```diff
@@ -181,21 +181,34 @@ dt_iop_colorspace_type_t default_colorspace(dt_iop_module_t *self,
   return IOP_CS_LAB;
 }
 
-int legacy_params(dt_iop_module_t *self, const void *const old_params, const int old_version, void *new_params,
-                  const int new_version)
+int legacy_params(dt_iop_module_t *self,
+                  const void *const old_params,
+                  const int old_version,
+                  void **new_params,
+                  int32_t *new_params_size,
+                  int *new_version)
 {
-  if(old_version == 1 && new_version == 3)
+  typedef struct dt_iop_colorbalance_params_v3_t
+  {
+    dt_iop_colorbalance_mode_t mode;
+    float lift[CHANNEL_SIZE], gamma[CHANNEL_SIZE], gain[CHANNEL_SIZE];
+    float saturation;
+    float contrast;
+    float grey;
+    float saturation_out;
+  } dt_iop_colorbalance_params_v3_t;
+
+  if(old_version == 1)
   {
     typedef struct dt_iop_colorbalance_params_v1_t
     {
       float lift[CHANNEL_SIZE], gamma[CHANNEL_SIZE], gain[CHANNEL_SIZE];
     } dt_iop_colorbalance_params_v1_t;
 
-    dt_iop_colorbalance_params_v1_t *o = (dt_iop_colorbalance_params_v1_t *)old_params;
-    dt_iop_colorbalance_params_t *n = (dt_iop_colorbalance_params_t *)new_params;
-    const dt_iop_colorbalance_params_t *const d = (dt_iop_colorbalance_params_t *)self->default_params;
-
-    *n = *d; // start with a fresh copy of default parameters
```

Comment:
There's a potential future inconsistency here if the introspection defaults are changed on the `dt_iop_colorbalance_params_t` declaration.  Eliminating the copying of the default parameters structure means that both the old_version==1 and old_version==2 paths need to be updated appropriately in that case.

Reply:
But colorbalance is a deprecated module so we should be safe :)